ngraph_bridge/ngraph_data_cache.h

```diff
@@ -0,0 +1,180 @@
+  std::pair<Status, ValueType> LookUpOrCreate(
+      KeyType key,
+      std::function<std::pair<Status, ValueType>()> callback_create_item,
+      std::function<void(ValueType)> callback_destroy_item);
+
+  // Overload for above function, which doesn't take callback for destroy item
+  std::pair<Status, ValueType> LookUpOrCreate(
+      KeyType key,
+      std::function<std::pair<Status, ValueType>()> callback_create_item);
+  void RemoveItem(KeyType key);
+  void RemoveItem(KeyType key,
+                  std::function<void(ValueType)> callback_destroy_item);
+  void RemoveAll(std::function<void(ValueType)> callback_destroy_item);
+
+ private:
+  std::unordered_map<KeyType, ValueType> m_ng_items_map;
+  std::deque<KeyType> m_lru;
+  int m_depth;
+  absl::Mutex m_mutex;
+
+  // Test class
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_SameKeyMultiThread_Test;
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_RemoveItemTest_Test;
+};
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::NgraphDataCache(int depth)
+    : m_depth(depth) {}
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::~NgraphDataCache() {
+  m_ng_items_map.clear();
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveItem(
+    KeyType key, std::function<void(ValueType)> callback_destroy_item) {
+  absl::MutexLock lock(&m_mutex);
+  if (m_ng_items_map.find(key) != m_ng_items_map.end()) {
+    callback_destroy_item(m_ng_items_map.at(key));
+    m_ng_items_map.erase(key);
+    m_lru.pop_back();
+  }
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveItem(KeyType key) {
+  RemoveItem(key, [](ValueType) {});
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveAll(
+    std::function<void(ValueType)> callback_destroy_item) {
+  absl::MutexLock lock(&m_mutex);
+  for (auto it = m_ng_items_map.begin(); it != m_ng_items_map.end(); it++) {
+    callback_destroy_item(it->second);
+  }
+  m_ng_items_map.erase(m_ng_items_map.begin(), m_ng_items_map.end());
+}
+
+template <typename KeyType, typename ValueType>
+std::pair<Status, ValueType>
+NgraphDataCache<KeyType, ValueType>::LookUpOrCreate(
+    KeyType key,
+    std::function<std::pair<Status, ValueType>()> callback_create_item,
+    std::function<void(ValueType)> callback_destroy_item) {
+  bool found_in_cache;
+  // look up in the cache
+  {
+    absl::MutexLock lock(&m_mutex);
+    auto it = m_ng_items_map.find(key);
+    found_in_cache = (it != m_ng_items_map.end());
+    if (found_in_cache) {
+      return std::make_pair(Status::OK(), m_ng_items_map.at(key));
+    }
+  }
+  // Item not found in cache, create item
+  ValueType item;
+  auto status_item_pair = callback_create_item();
```

Comment:
try + catch the invocation of callbacks . we can throw errors like "Failed to create an item. Invalid Callback to Create"

From here:
https://en.cppreference.com/w/cpp/utility/functional/function
The stored callable object is called the target of std::function. If a std::function contains no target, it is called empty. Invoking the target of an empty std::function results in std::bad_function_call exception being thrown.

Reply:
Done

Reply:
Done.